source/DistanceMap.cpp

```diff
@@ -70,6 +70,20 @@ DistanceMap::DistanceMap(const Ship &ship, const System *destination)
 
 
 
+// Calculate the path for the given ship to travel between two system. The
+// ship will use a jump drive or hyperdrive depending on what it has. The
+// pathfinding will stop once a path to the destination is found.
+DistanceMap::DistanceMap(const Ship &ship, const System *source, const System *destination)
```

Comment:
Rather than essentially duplicate the above constructor, I would extend it with a default nullptr parameter:
```
DistanceMap::DistanceMap(const Ship &ship, const System *destination, const System *source)
    : source(source ? source : ship.GetSystem())
```

Reply:
I rather not, since DistanceMap constructors are the only way to configure the class.
I want a clear distinction between "from ship to system" and "from X to Y with this ship".

I want to be able to pass a nullptr source since it can simplify logic that uses DistanceMap.
I have the other constructor for when I don't care about the source.

Reply:
As far as I can tell, you aren't able to pass `nullptr` to this constructor with this PR, since `from` can only be a system from the player's travel plan (i.e. not `nullptr`), or the ship's target system (which is verified not to be `nullptr`) before the constructor is called.

Reply:
Regardless of the current use, having an different source when I provide my own (via variable) would be counter intuitive.
DistanceMap already has counter intuitive code (search direction vs variable names) which frustrated me and such a constructor would be another one. For me, clearer purpose is better than less functions.

Reply:
I think most are in agreement that clear functions are preferred to obtuse code :+1: 

To me it makes sense that if `source` is passed, it should be used if possible, and if `source` is not passed (or is given but doesn't exist), but a ship was also passed, then `source` should be that ship's current system.

You mentioned logic that is simplified if DistanceMap is given a nullptr `source` (e.g. aborted object instantiation). Can you share some uses of this?

Reply:
One case would be dynamic mission construction. Something like:
```
const System *from = ???;// a remote system according to some location filter or nullptr if noone match
const System *to = ???;// another remote system according to some location filter or nullptr if noone match
DistanceMap route(ship, from, to);// here ship might be the player flagship or a npc ship (or both with another DistanceMap)
if(route.HasRoute())
  ???; // set up "events" along the route
else
 ???; // can't construct the mission right now, abort
```
No need to test if it found systems matching the location filters, `DistanceMap` already checks for `nullptr` and `DistanceMap::HasRoute()` gives the expected "no route" result.

Reply:
That's a great idea, and you should pursue it. Probably better to save the constructor for that PR though.